### PR TITLE
[#9120] Deactivate all sessions in deleted courses

### LINKED ISSUE #9120: Deactivate all sessions in deleted courses
**Environment:**
`master` branch at commit https://github.com/TEAMMATES/teammates/commit/51e4279ad8e098ed25d01b5a332a7d5b514d858a

**Description:**
Similar to #9109 . When soft-deleting a course, the sessions inside are still active. Need to check whether the corresponding course has been deleted before sending emails.

### PR DESCRIPTION
**Pull request by dalessr**
Fixes #9120

### CODE REVIEW

**Comment by xpdavid** [comment]
Actually, I was thinking about why we don't do cascade soft-deleted for the corresponding sessions when a course is soft-deleted.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 1]
```diff
@@ -191,6 +193,7 @@ private void testGetFeedbackSessionsClosingWithinTimeLimit() throws Exception {
         session.setEndTime(TimeHelper.getInstantDaysOffsetFromNow(1));
         ThreadHelper.waitBriefly(); // this one is correctly used
         fsLogic.createFeedbackSession(session);
+        coursesLogic.createCourse(session.getCourseId(), "Test Course", "UTC");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: The test is for?

Comment (dalessr) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: When user adds a session from the `Action` layer (`InstructorFeedbackAddAction`), it will first check & ensure that the session belongs to an existing course. However in this test case, the course doesn't exist in the datastore. It simply adds a single session to test the behaviors inside. Thus I'm adding the course here so that it can check whether the course is deleted in `FeedbackSessionsLogic`. Then we can decide whether to send email for the session.

**Comment by dalessr** [comment]
> We should add test for the introduced behaviours.

Yup will add/modify the tests to `FeedbackSessionsLogicTest`

> Actually, I was thinking about why we don't do cascade soft-deleted for the corresponding sessions when a course is soft-deleted.

I discussed about this with Prof @damithc before. It is not very ideal. Consider this case:
1. Create a course and add some sessions inside.
2. Delete a session inside the course. Then the session goes to Recycle Bin.
3. Delete the course. Then the course goes to Recycle Bin. Same as all the sessions inside.
4. Restore the course. Then all the sessions are also restored, including the session that was previously deleted. If that session is open/active, then wrong emails might be sent. And it's also inconvenient for user to identify & distinguish.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 2]
```diff
@@ -222,33 +228,58 @@ private void testGetFeedbackSessionsWhichNeedOpenMailsToBeSent() throws Exceptio
...
         session.setEndTime(TimeHelperExtension.getInstantHoursOffsetFromNow(-1));
         fsLogic.updateFeedbackSession(session);
 
-        sessionList = fsLogic
-                .getFeedbackSessionsWhichNeedOpenEmailsToBeSent();
+        sessionList = fsLogic.getFeedbackSessionsWhichNeedOpenEmailsToBeSent();
+
+        assertEquals(0, sessionList.size());
+
+        ______TS("case : 1 open session in soft-deleted course with mail unsent");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: should use `recycle bin` rather than `soft-deleted` in logic layer right

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 3]
```diff
@@ -1467,12 +1467,11 @@ public void unpublishFeedbackSession(FeedbackSessionAttributes sessionToUnpublis
      */
     public List<FeedbackSessionAttributes> getFeedbackSessionsClosedWithinThePastHour() {
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Test for this method?

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java (hunk 2 above)
Comment (xpdavid) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: We need to make them consistent. My suggestion is that in logic layer, we should refrain from mentioning the term 'soft-deleted'.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 4]
```diff
@@ -178,42 +181,78 @@ private void testIsFeedbackSessionHasQuestionForStudents() throws Exception {
...
+
+        assertEquals(0, sessionList.size());
+
+        ______TS("case : 1 closed session in undeleted course within the past hour");
+        FeedbackSessionAttributes session = getNewFeedbackSession();
+        session.setTimeZone(ZoneId.of("UTC"));
+        session.setSessionVisibleFromTime(TimeHelper.getInstantDaysOffsetFromNow(-1));
+        session.setStartTime(TimeHelper.getInstantDaysOffsetFromNow(-1));
+        session.setEndTime(TimeHelperExtension.getInstantHoursOffsetFromNow(-1));
+        ThreadHelper.waitBriefly(); // this one is correctly used
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Why is this necessary (I know other tests also have)? In addition, the comment is useless.

Comment (xpdavid) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: I guess here we are trying to test the grace period?

```java
    public boolean isClosedWithinPastHour() {
        Instant now = Instant.now();
        Instant given = endTime.plus(gracePeriod);
        return given.isBefore(now) && Duration.between(given, now).compareTo(Duration.ofHours(1)) < 0;
    }
```

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 5]
```diff
@@ -222,33 +261,58 @@ private void testGetFeedbackSessionsWhichNeedOpenMailsToBeSent() throws Exceptio
...
-                .getFeedbackSessionsWhichNeedOpenEmailsToBeSent();
+        sessionList = fsLogic.getFeedbackSessionsWhichNeedOpenEmailsToBeSent();
+
         assertEquals(0, sessionList.size());
 
-        //delete the newly added session as removeAndRestoreTypicalDataInDatastore()
-        //wont do it
-        fsLogic.deleteFeedbackSessionCascade(session.getFeedbackSessionName(),
-                session.getCourseId());
+        ______TS("case : 1 open session in deleted course with mail unsent");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: The session is not open right?

In L279, you set the `endTime` to one hour ago and never set back in this test case.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java (hunk 4 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Hmm... ya should be. I think this method is not necessary here.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java (hunk 5 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Opps!

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 6]
```diff
@@ -178,42 +181,77 @@ private void testIsFeedbackSessionHasQuestionForStudents() throws Exception {
...
+        List<FeedbackSessionAttributes> sessionList = fsLogic.getFeedbackSessionsClosedWithinThePastHour();
+
+        assertEquals(0, sessionList.size());
+
+        ______TS("case : 1 closed session in undeleted course within the past hour");
+        FeedbackSessionAttributes session = getNewFeedbackSession();
+        session.setTimeZone(ZoneId.of("UTC"));
+        session.setSessionVisibleFromTime(TimeHelper.getInstantDaysOffsetFromNow(-1));
+        session.setStartTime(TimeHelper.getInstantDaysOffsetFromNow(-1));
+        session.setEndTime(TimeHelperExtension.getInstantHoursOffsetFromNow(-1));
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: I would say using `getInstantMinutesOffsetFromNow([10..50])` is more straightforward. If we offset one hour. By right, it is not session `ClosedWithinThePastHour`. It is a bit confusing here as get `gracePeriod` involved.

**Comment by xpdavid** [comment]
Regarding the question:

> Restore the course. Then all the sessions are also restored, including the session that was previously deleted. If that session is still open/active, then wrong emails might be sent. And it's also inconvenient for user to identify & distinguish.

We don't necessary need to restored all sessions -> restore no session and leave the user to choose what session to restore. Even if we want to distinguished between what is originally in the recycle bin and what is not, we can do this by comparing the `deleteTime`. Conceptually, cascade deleted course should have the same deleted time as the corresponding session.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java (hunk 6 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: Will set as `-59` due to BVA?

**Comment by dalessr** [comment]
> Even if we want to distinguished between what is originally in the recycle bin and what is not, we can do this by comparing the deleteTime. Conceptually, cascade deleted course should have the same deleted time as the corresponding session.

Ohh good point! Never thought of that before. But for now I think we can just keep the original design first as it's not actually a serious issue. Let's merge & deploy this fix as early as possible :)
